core/lib/makeswift/components/product-detail/client.tsx

```diff
@@ -0,0 +1,123 @@
+'use client';
+
+import React, {
+  type ComponentPropsWithoutRef,
+  createContext,
+  forwardRef,
+  type PropsWithChildren,
+  type ReactNode,
+  type Ref,
+  useCallback,
+  useContext,
+} from 'react';
+
+import { Stream, type Streamable } from '@/vibes/soul/lib/streamable';
+import { ProductDetail, type ProductDetailProduct } from '@/vibes/soul/sections/product-detail';
+import { mergeSections } from '~/lib/makeswift/utils/merge-sections';
+
+type VibesProductDetailProps = ComponentPropsWithoutRef<typeof ProductDetail>;
+export type ProductDetail = ProductDetailProduct & {
+  plainTextDescription?: string;
+};
+
+export type Props = Omit<VibesProductDetailProps, 'product'> & {
+  product: Streamable<ProductDetail>;
+};
+
+const PropsContext = createContext<Props | null>(null);
+
+export const PropsContextProvider = ({ value, children }: PropsWithChildren<{ value: Props }>) => (
+  <PropsContext.Provider value={value}>{children}</PropsContext.Provider>
+);
+
+export const DescriptionSource = {
+  CatalogPlainText: 'CatalogPlainText',
+  CatalogRichText: 'CatalogRichText',
+  Custom: 'Custom',
+} as const;
+
+type DescriptionSource = (typeof DescriptionSource)[keyof typeof DescriptionSource];
+
+interface EditableProps {
+  summaryText: string | undefined;
+  description: { source: DescriptionSource; slot: ReactNode };
+  accordions: Exclude<Awaited<ProductDetailProduct['accordions']>, undefined>;
+}
+
+const ProductDetailImpl = ({
+  summaryText,
+  description,
+  accordions,
+  product: streamableProduct,
+  ...props
+}: Props & EditableProps) => {
+  const getProductDescription = useCallback(
+    (product: ProductDetail): ProductDetail['description'] => {
+      switch (description.source) {
+        case DescriptionSource.CatalogPlainText:
+          return product.plainTextDescription;
+
+        case DescriptionSource.CatalogRichText:
+          return product.description;
+
+        case DescriptionSource.Custom:
+          return description.slot;
+      }
+    },
+    [description.source, description.slot],
+  );
+
+  const getProductAccordions = useCallback(
+    (
+      productAccordions: Awaited<ProductDetail['accordions']>,
+    ): Awaited<ProductDetail['accordions']> =>
+      productAccordions != null
+        ? mergeSections(productAccordions, accordions, (left, right) => ({
+            ...left,
+            content: right.content,
+          }))
+        : undefined,
+    [accordions],
+  );
+
+  return (
+    <Stream value={streamableProduct}>
+      {(product) => (
+        <Stream value={product.accordions}>
+          {(productAccordions) => (
```

Comment:
@migueloller Nested streamables still cause a bit of complexity here.

Reply:
All good. Don't we need to provide appropriate fallbacks here so that we don't see FOUC?

Reply:
It didn't seem to make a difference, but I've added them anyway.